### PR TITLE
Minor cleanup for handlers/watch

### PR DESCRIPTION
**Pull request by linxiulei**
#### What type of PR is this?

/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

Cleanups for #120902

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go [hunk 1]
```diff
@@ -258,17 +258,12 @@ func ListResource(r rest.Lister, rw rest.Watcher, scope *RequestScope, forceWatc
 				timeout = time.Duration(float64(minRequestTimeout) * (rand.Float64() + 1.0))
 			}
 			klog.V(3).InfoS("Starting watch", "path", req.URL.Path, "resourceVersion", opts.ResourceVersion, "labels", opts.LabelSelector, "fields", opts.FieldSelector, "timeout", timeout)
-			ctx, cancel := context.WithTimeout(ctx, timeout)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go:261]: Please move the first commit to a separate PR - I would definitely want more pairs of eye to look at it, we had a number of issues with timeouts in the past...

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go [hunk 2]
```diff
@@ -64,7 +64,12 @@ func (w *realTimeoutFactory) TimeoutCh() (<-chan time.Time, func() bool) {
 // serveWatch will serve a watch response.
 // TODO: the functionality in this method and in WatchServer.Serve is not cleanly decoupled.
 func serveWatch(watcher watch.Interface, scope *RequestScope, mediaTypeOptions negotiation.MediaTypeOptions, req *http.Request, w http.ResponseWriter, timeout time.Duration, metricsScope string) {
-	defer watcher.Stop()
+	closeWatch := true
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:67]: Let's not move this part here yet - if for whatever reason we decide to not proceed with the main change, this isn't desired change.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go [hunk 3]
```diff
@@ -187,11 +194,23 @@ type WatchServer struct {
 // ServeHTTP serves a series of encoded events via HTTP with Transfer-Encoding: chunked
 // or over a websocket connection.
 func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	returnEarly := true
+	defer func() {
+		if returnEarly {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:199]: For now [given we don't offload to a separate goroutine]:
- let's not move watch ownership to here [as asked in the comment above]
- then for MemoryAllocator you can simply fo:
```
defer func() {
  if s.MemoryAllocator != nil {
    runtime.AllocatorPool.Put(s.MemoryAllocator)
  }
}()
```

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go [hunk 4]
```diff
@@ -212,63 +231,79 @@ func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+		serveFunc()
+	})
 }
 
 // HandleWS implements a websocket handler.
 func (s *WatchServer) HandleWS(ws *websocket.Conn) {
 	defer ws.Close()
 	done := make(chan struct{})
+	// ensure the connection times out
+	timeoutCh, cleanup := s.TimeoutFactory.TimeoutCh()
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:305]: Same for this one - move it together to the PR that will only be touching timeouts.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go (hunk 2 above)
Comment (linxiulei) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go]: Done

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go (hunk 3 above)
Comment (linxiulei) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go]: Done

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go (hunk 4 above)
Comment (linxiulei) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go]: Done in #122060

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go [hunk 5]
```diff
@@ -222,47 +230,53 @@ func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+					flusher.Flush()
+				}
+				if isWatchListLatencyRecordingRequired {
+					metrics.RecordWatchListLatency(req.Context(), s.Scope.Resource, s.metricsScope)
+				}
 			}
 		}
 	}
+
+	metrics.RecordLongRunning(req, requestInfo, metrics.APIServerComponent, func() {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:277]: nit:
```
metrics.RecordLongRunning(req, requestInfo, metrics.APIServerComponent, serveFunc)
```

**Review by wojtek-t — COMMENTED** [review]
Also - please squash commits

**Comment by linxiulei** [comment]
> Also - please squash commits

Done and addressed all comments.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go [hunk 6]
```diff
@@ -222,47 +230,51 @@ func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+					flusher.Flush()
+				}
+				if isWatchListLatencyRecordingRequired {
+					metrics.RecordWatchListLatency(req.Context(), s.Scope.Resource, s.metricsScope)
+				}
 			}
 		}
 	}
+
+	metrics.RecordLongRunning(req, requestInfo, metrics.APIServerComponent, serveFunc)
```
[the remark this answers is not in the document]
Comment (dgrisonnet) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:277]: I also think that the new approach is fine.

In theory, accounting for the connection setup would make the measurement more accurate, but in practice, it shouldn't make much difference.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go [hunk 7]
```diff
@@ -222,47 +230,51 @@ func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
-			// available apiserver instance(s).
-			return
-		case <-done:
-			return
-		case <-timeoutCh:
-			return
-		case event, ok := <-ch:
-			if !ok {
-				// End of results.
+	serveFunc := func() {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:233]: I decided to clean this up a little bit in https://github.com/kubernetes/kubernetes/pull/122115
Let's wait for that PR and adjust on top of that one - I hope it should help a bit with this refactoring.

Comment (linxiulei) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go]: @wojtek-t I rebased on top of https://github.com/kubernetes/kubernetes/pull/122115, which made this refactoring much simpler so we might not need #122060. PTAL.

**Comment by wojtek-t** [comment]
This looks much better on top of my PR and is much easier to follow - thanks!

/lgtm
/approve
